### DISCUSSION TITLE
Is there any way to send "exit" programmatically?

### DISCUSSION THREAD

**Discussion by monteduro** [microsoft/autogen#2087] [Q&A]
I'm using Autogen via WebSocket with the new initiate_chats() and a human_input_mode set to TERMINATE. That's because I need frequent real user input.

Everything looks okay via Terminal, especially when a chat completes its work and it's time to move to the next one: I just have to type 'exit' or press enter.

BUT if I send 'exit' or an empty string via socket (using in pytnon user.send(message, agent_talking)), it will start a loop where the agent asks if it can do something else and the user keeps saying 'exit'.

I've read ALL the code and I didn't find a single line that would allow me to stop a chat programmatically.

Am I dumb? Probably. So here I'm asking for help.

**Comment by sonichi** [comment]
Try send "TERMINATE" when using `user.send()`. "exit" is only processed inside `check_termination_and_get_human_input`

**Reply by monteduro** [reply]
<img width="589" alt="Screenshot 2024-03-20 alle 06 47 04" src="https://github.com/microsoft/autogen/assets/5638815/74859932-246b-4016-8ced-99c51c6bee03">

Nothing happened. Even if I set this to the mail assistant:
```
human_input_mode="TERMINATE",  # ask human for input at each step
is_termination_msg=is_termination_msg,
```

The only way to proceed further is to type enter or exit via terminal. 

I'm wondering if autogen is the right tool for handling complex tasks while maintaining constant interaction with the real user. In my test, I'm trying to:

1. Have the user define the type of email to send. The language model queries a function with the available email types and presents the choices to the user.
2. When the user makes a selection, it should proceed to the next chat, where the agent retrieves the necessary information from the user (or from some functions) based on the 'required_informations' of the email.
3. The last chat is with the agent tasked with writing the email using all the collected information and structure from the previous chat.

Via the terminal, it's not working badly, although some functions haven't worked well. However, in attempting to interact with the frontend, I've become completely stuck due to this problem.

**Comment by monteduro** [comment]
Ugly but functional awaiting #1551. I have overridden the get_human_input method to support streaming with SocketIO. Additionally, I have set a global variable next_chat_trigger which is set to True when the LLM message ends with "COMPLETED", thus forcing an 'exit' and moving to the next sequential chat.

```python
    def get_human_input(self, prompt: str) -> str:
        global message_chat
        global next_chat_trigger
        
        if ( next_chat_trigger ):
            print('next_chat_trigger')
            next_chat_trigger = False
            message_chat = None
            return 'exit'
        else:
            socketio.emit('message', {"sender": agent_talking.name,
                    "content": prompt}, to = session['current_session_id'])

        while not message_chat:
            pass
        
        reply = message_chat
        message_chat = None

        self._human_input.append(reply)
        socketio.emit('message', {"sender": user.name,
                    "content": reply}, to = session['current_session_id'])
        
        return reply
        
```

**Comment by Pankaj-Samyak** [comment]
@monteduro Did you find any way yet?
